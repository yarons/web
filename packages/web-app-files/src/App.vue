<template>
  <main id="files" class="oc-flex oc-height-1-1">
    <div
      ref="filesListWrapper"
      tabindex="-1"
      class="files-list-wrapper oc-width-expand"
      @dragover="$_ocApp_dragOver"
    >
      <router-view id="files-view" />
    </div>
    <side-bar
      v-if="showSidebar"
      id="files-sidebar"
      ref="filesSidebar"
      tabindex="-1"
      class="oc-width-1-1 oc-width-2-5@m oc-width-1-4@xl"
      :sidebar-active-panel="sidebarActivePanel"
      @beforeDestroy="focusSideBar"
      @mounted="focusSideBar"
      @fileChanged="focusSideBar"
      @selectPanel="setActiveSidebarPanel"
      @close="closeSidebar"
    />
  </main>
</template>
<script lang="ts">
import Mixins from './mixins'
import { mapActions, mapGetters, mapState } from 'vuex'
import SideBar from './components/SideBar/SideBar.vue'
import { defineComponent } from '@vue/composition-api'

export default defineComponent({
  components: {
    SideBar
  },
  mixins: [Mixins],
  computed: {
    ...mapGetters('Files', ['dropzone']),
    ...mapState('Files/sidebar', {
      sidebarClosed: 'closed',
      sidebarActivePanel: 'activePanel'
    }),

    showSidebar() {
      return !this.sidebarClosed
    }
  },
  watch: {
    $route: {
      handler: function (to, from) {
        this.resetFileSelection()
        if (from?.name !== to.name) {
          this.closeSidebar()
        }
      }
    }
  },
  created() {
    this.$root.$on('upload-end', () => {
      this.delayForScreenreader(() => this.$refs.filesListWrapper.focus())
    })
  },

  methods: {
    ...mapActions('Files', ['dragOver', 'resetFileSelection']),
    ...mapActions('Files/sidebar', {
      closeSidebar: 'close',
      setActiveSidebarPanel: 'setActivePanel'
    }),
    ...mapActions(['showMessage']),

    focusSideBar(component, event) {
      this.focus({
        from: document.activeElement,
        to: this.$refs.filesSidebar?.$el,
        revert: event === 'beforeDestroy'
      })
    },
    $_ocApp_dragOver(event) {
      const hasfileInEvent = (event.dataTransfer.types || []).some((e) => e === 'Files')
      this.dragOver(hasfileInEvent)
    }
  }
})
</script>

<style lang="scss" scoped>
main {
  max-height: 100%;
}

.files-list-wrapper {
  overflow-y: auto;
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: max-content max-content 1fr;
  gap: 0 0;
  grid-template-areas:
    'header'
    'upload'
    'main';

  &:focus {
    outline: none;
  }
}

#files-sidebar {
  position: relative;
  overflow: hidden;
}

#files-view {
  grid-area: main;
  z-index: 0;
}

#files-upload-progress {
  grid-area: upload;
}
</style>
