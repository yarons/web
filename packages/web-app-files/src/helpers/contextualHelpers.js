// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}
export const empty = {
  text: $gettext(''),
  list: [$gettext(''), $gettext(''), $gettext('')]
}
export const shareInviteCollaboratorHelp = {
  text: $gettext('Invite persons or groups to access this file or folder.'),
  list: [
    $gettext('Enter a name or group to share this item'),
    $gettext(
      'If you share a folder,  all of its contents and subfolders will be shared with the entered persons or groups'
    ),
    $gettext('Invited persons or groups will be notified via e-mail or in-app notification'),
    $gettext('Invited persons can not see who else has access')
  ]
}
export const shareSpaceAddMemberHelp = {
  text: $gettext('Add persons or groups as members to this Space. Members are allowed to:'),
  list: [
    $gettext('see who else is member of this space'),
    $gettext('view all files in this space'),
    $gettext('(if permitted) edit or delete files in this Space'),
    $gettext('(if permitted) share folders of this Space with non-Members'),
    $gettext('see with whom a folder is shared')
  ],
  endText: 'Members can only be added or removed by a Space Manager.'
}
export const shareViaLinkHelp = {
  text: $gettext('Share a file or folder by link. Choose how access is granted:'),
  list: [
    $gettext(
      'Only invited people can access: Only people from the list "Invited people" can access. If there is no list, no people are invited yet.'
    ),
    $gettext(
      'Everyone with the link: Everyone with the link can access. Note: If you share this link with people from the list "Invited people", they need to login-in so that their individual assigned permissions can take effect. If they are not logged-in, the permissions of the link take effect.'
    )
  ]
}
export const shareQuickLinkHelp = {
  text: $gettext('The Quick link is the link that is always copied if you'),
  list: [
    $gettext('right click on a file and choose "Get link" or'),
    $gettext('click on the quickaction "Get link"')
  ]
}
export const shareAddLinkHelp = {
  text: $gettext('Add and manage multiple links for the same file or folder.')
}
