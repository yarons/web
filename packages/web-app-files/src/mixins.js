import { DateTime } from 'luxon'

export default {
  filters: {
    roundNumber(int) {
      return parseInt(int.toFixed(0))
    }
  },
  methods: {
    formDateFromJSDate(date, format = DateTime.DATETIME_FULL) {
      return DateTime.fromJSDate(date).setLocale(this.$language.current).toLocaleString(format)
    },
    formDateFromHTTP(date, format = DateTime.DATETIME_FULL) {
      return DateTime.fromHTTP(date).setLocale(this.$language.current).toLocaleString(format)
    },
    formDateFromISO(date, format = DateTime.DATETIME_FULL) {
      return DateTime.fromISO(date).setLocale(this.$language.current).toLocaleString(format)
    },
    formDateFromRFC(date, format = DateTime.DATETIME_FULL) {
      return DateTime.fromRFC2822(date).setLocale(this.$language.current).toLocaleString(format)
    },
    formRelativeDateFromJSDate(date) {
      return DateTime.fromJSDate(date).setLocale(this.$language.current).toRelative()
    },
    formRelativeDateFromHTTP(date) {
      return DateTime.fromHTTP(date).setLocale(this.$language.current).toRelative()
    },
    formRelativeDateFromISO(date) {
      return DateTime.fromISO(date).setLocale(this.$language.current).toRelative()
    },
    formRelativeDateFromRFC(date) {
      return DateTime.fromRFC2822(date).setLocale(this.$language.current).toRelative()
    },
    delayForScreenreader(func, delay = 500) {
      /*
       * Delay for screen readers Virtual buffers
       */
      setTimeout(() => func(), delay)
    }
  }
}
