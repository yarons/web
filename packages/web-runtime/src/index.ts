import {
  DesignSystem as designSystem,
  pages,
  translations,
  supportedLanguages,
  store,
  Vue
} from './defaults'

import { router } from './router'

import {
  requestConfiguration,
  announceApplications,
  announceClient,
  announceDefaults,
  announceClientService,
  announceStore,
  announceTheme,
  announceTranslations,
  announceVersions,
  applicationStore,
  announceUppyService
} from './container'

export const bootstrap = async (configurationPath: string): Promise<void> => {
  const runtimeConfiguration = await requestConfiguration(configurationPath)
  announceClientService({ vue: Vue, runtimeConfiguration })
  announceUppyService({ vue: Vue })
  await announceClient(runtimeConfiguration)
  await announceStore({ vue: Vue, store, runtimeConfiguration })
  await announceApplications({
    runtimeConfiguration,
    store,
    supportedLanguages,
    router,
    translations
  })
  announceTranslations({ vue: Vue, supportedLanguages, translations })
  await announceTheme({ store, vue: Vue, designSystem, runtimeConfiguration })
  announceDefaults({ store, router })
}

export const renderSuccess = (): void => {
  announceVersions({ store })
  const applications = Array.from(applicationStore.values())
  const instance = new Vue({
    el: '#owncloud',
    store,
    router,
    render: (h) => h(pages.success)
  })

  instance.$once('mounted', () => {
    applications.forEach((application) => application.mounted(instance))
  })

  store.watch(
    (state, getters) => getters.isUserReady,
    (newValue, oldValue) => {
      if (!newValue || newValue === oldValue) {
        return
      }
      applications.forEach((application) => application.userReady(instance))
    },
    {
      immediate: true
    }
  )
}

export const renderFailure = async (err: Error): Promise<void> => {
  announceVersions({ store })
  await announceTranslations({ vue: Vue, supportedLanguages, translations })
  await announceTheme({ store, vue: Vue, designSystem })
  console.error(err)
  new Vue({
    el: '#owncloud',
    store,
    render: (h) => h(pages.failure)
  })
}
