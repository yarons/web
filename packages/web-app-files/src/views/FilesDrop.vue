<template>
  <div id="files-drop-container" class="oc-height-1-1 oc-flex oc-flex-column oc-flex-between">
    <h1 class="oc-invisible-sr">{{ pageTitle }}</h1>
    <div class="oc-p oc-height-1-1">
      <div v-if="loading" key="loading-drop" class="oc-flex oc-flex-column oc-flex-middle">
        <h2 class="oc-login-card-title">
          <translate>Loading public linkâ€¦</translate>
        </h2>
        <oc-spinner :aria-hidden="true" />
      </div>
      <div v-else key="loaded-drop" class="oc-flex oc-flex-column oc-flex-middle oc-height-1-1">
        <div class="oc-text-center oc-width-1-1 oc-width-xxlarge@m">
          <h2 v-text="title" />
          <!-- needs uppy dropzone -->
          <div id="files-drop-zone" class="uk-flex uk-flex-middle uk-flex-center uk-placeholder">
            <oc-icon name="file_upload" />
            <translate>Drop files here to upload or click to select file</translate>
          </div>
          <div id="previews" hidden />
        </div>
        <div
          class="oc-flex oc-flex-center oc-overflow-auto oc-width-1-1 oc-mt"
          :class="{ 'files-empty': !getUploadedFiles }"
        >
          <oc-table-simple v-if="getUploadedFiles" class="oc-width-1-1 oc-width-xxlarge@m">
            <oc-tbody>
              <oc-tr v-for="(file, key) in getUploadedFiles" :key="key">
                <oc-td class="oc-pl-rm" v-text="file.name" />
                <oc-td width="shrink" class="oc-text-nowrap oc-text-muted">
                  <oc-resource-size :size="file.size" />
                </oc-td>
                <oc-td width="shrink" class="oc-pr-rm">
                  <oc-icon
                    v-if="file.status === 'done'"
                    name="checkbox-circle"
                    size="small"
                    variation="success"
                  />
                  <oc-icon
                    v-if="file.status === 'error'"
                    name="information"
                    size="small"
                    variation="danger"
                  />
                  <oc-spinner
                    v-if="file.status === 'uploading' || file.status === 'init'"
                    size="small"
                    :aria-label="$_ocUploadingFileMessage(file.name)"
                  />
                </oc-td>
              </oc-tr>
            </oc-tbody>
          </oc-table-simple>
        </div>
        <div v-if="errorMessage" class="oc-text-center">
          <h2>
            <translate>An error occurred while loading the public link</translate>
          </h2>
          <p class="oc-m-rm" v-text="errorMessage" />
        </div>
      </div>
    </div>
    <div class="oc-text-center">
      <p v-text="configuration.currentTheme.general.slogan" />
    </div>
  </div>
</template>

<script>
import { mapGetters } from 'vuex'
import { DavProperties, DavProperty } from 'web-pkg/src/constants'
import { linkRoleUploaderFolder } from '../helpers/share'
import { createLocationOperations, createLocationPublic } from '../router'

import Uppy from '@uppy/core'
import Tus from '@uppy/tus'
import XHRUpload from '@uppy/xhr-upload'
import StatusBar from '@uppy/status-bar'
import DropTarget from '@uppy/drop-target'

export default {
  data() {
    return {
      loading: true,
      errorMessage: null,
      uploadedFiles: [],
      uploadedFilesChangeTracker: 0
    }
  },
  computed: {
    ...mapGetters(['capabilities', 'configuration', 'newFileHandlers', 'user']),
    ...mapGetters('Files', ['currentFolder', 'publicLinkPassword']),
    pageTitle() {
      return this.$gettext(this.$route.meta.title)
    },
    publicLinkToken() {
      return this.$route.params.token
    },
    title() {
      // share might not be loaded
      if (this.share) {
        const translated = this.$gettext('%{owner} shared this folder with you for uploading')
        return this.$gettextInterpolate(
          translated,
          { owner: this.share.getProperty(this.$client.publicFiles.PUBLIC_LINK_SHARE_OWNER) },
          true
        )
      }
      return ''
    },
    url() {
      return this.$client.publicFiles.getFileUrl(this.publicLinkToken) + '/'
    },
    getUploadedFiles() {
      return this.uploadedFilesChangeTracker && this.uploadedFiles.values()
    }
  },
  mounted() {
    this.resolvePublicLink()

    const client = this.$client
    const uppyHeaders = client.helpers.buildHeaders()

    // might make sense to initialize an Uppy instance in the runtime?
    // TODO: set debug to false
    const uppy = new Uppy({ debug: true, autoProceed: true })

    // TODO: Build headers differently in public context?
    // if (this.publicPage()) { ... maybe use this.headers to obtain token ... }

    // TODO: What about flaky capability loading and its implications?
    // if (this.capabilities?.files.tus_support?.max_chunk_size > 0) {
    //   const chunkSize =
    //     this.capabilities.files.tus_support.max_chunk_size > 0 &&
    //     this.configuration.uploadChunkSize !== Infinity
    //       ? Math.max(
    //           this.capabilities.files.tus_support.max_chunk_size,
    //           this.configuration.uploadChunkSize
    //         )
    //       : this.configuration.uploadChunkSize

    delete uppyHeaders['OCS-APIREQUEST']

    uppy.use(Tus, {
      endpoint: this.url,
      headers: uppyHeaders,
      chunkSize: this.configuration.uploadChunkSize,
      removeFingerprintOnSuccess: true,
      overridePatchMethod: '',
      retryDelays: [0, 3000, 5000, 10000, 20000]
    })
    // } else {
    // uppy.use(XHRUpload, {
    //   endpoint: this.url,
    //   method: 'put',
    //   headers: uppyHeaders
    // })
    // }

    // // upload via drag&drop handling
    uppy.use(DropTarget, {
      target: '#files-drop-zone'
    })

    // upload button handling (files & folders separately)
    // doesn't recognize elements yet since they're tippy children, maybe use $refs?
    const uploadInputTarget = document.getElementById('#files-drop-zone')

    uploadInputTarget.addEventListener('change', (event) => {
      const files = Array.from(event.target.files)

      files.forEach((file) => {
        try {
          console.log('beginning upload for file:', file)
          uppy.addFile({
            source: 'file input',
            name: file.name,
            type: file.type,
            data: file
          })
        } catch (err) {
          console.error('error upload file:', file)
          if (err.isRestriction) {
            // handle restrictions
            console.log('Restriction error:', err)
          } else {
            // handle other errors
            console.error(err)
          }
        }
      })
    })

    // uppy.use(StatusBar, {
    //   id: 'StatusBar',
    //   target: '#files-app-bar',
    //   hideAfterFinish: true,
    //   showProgressDetails: true,
    //   hideUploadButton: false,
    //   hideRetryButton: false,
    //   hidePauseResumeButton: false,
    //   hideCancelButton: false,
    //   doneButtonHandler: null,
    //   locale: {
    //     // TODO: Uppy l10n research
    //   }
    // })

    uppy.on('upload-error', (file, error, response) => {
      console.log('error with file:', file.id)
      console.log('error message:', error)
      this.onFileError(error.toString())
    })

    uppy.on('file-removed', () => {
      uploadInputTarget.forEach((item) => {
        item.value = null
      })
    })

    uppy.on('complete', (result) => {
      result.successful.forEach((file) => {
        this.uploadedFiles.push(file)
      })
      uploadInputTarget.forEach((item) => {
        item.value = null
      })

      console.log('successful files:', result.successful)
      console.log('failed files:', result.failed)
    })
  },
  methods: {
    resolvePublicLink() {
      this.loading = true

      this.$client.publicFiles
        .list(this.publicLinkToken, this.publicLinkPassword, DavProperties.PublicLink, '0')
        .then((files) => {
          // Redirect to files list if the link doesn't have role "uploader"
          const sharePermissions = parseInt(files[0].getProperty(DavProperty.PublicLinkPermission))
          if (linkRoleUploaderFolder.bitmask(false) !== sharePermissions) {
            this.$router.replace(
              createLocationPublic('files-public-files', {
                params: { item: this.publicLinkToken }
              })
            )
            return
          }
          this.share = files[0]
        })
        .catch((error) => {
          // likely missing password, redirect to public link password prompt
          if (error.statusCode === 401) {
            this.$router.push(
              createLocationOperations('files-operations-resolver-public-link', {
                params: {
                  token: this.publicLinkToken
                }
              })
            )
            return
          }
          this.errorMessage = error
        })
        .finally(() => {
          this.loading = false
        })
    },
    dropZoneFileAdded(event) {
      const uploadId = event.upload.uuid
      this.uploadedFilesChangeTracker++
      this.uploadedFiles.set(uploadId, { name: event.name, size: event.size, status: 'init' })

      this.uploadQueue
        .add(() =>
          this.$client.publicFiles.putFileContents(
            this.publicLinkToken,
            event.name,
            this.publicLinkPassword,
            event,
            {
              // automatically rename in case of duplicates
              headers: { 'OC-Autorename': 1 },
              onProgress: (progressEvent) => {
                this.uploadedFilesChangeTracker++
                this.uploadedFiles.set(uploadId, {
                  name: event.name,
                  size: event.size,
                  status: 'uploading'
                })
              }
            }
          )
        )
        .then((e) => {
          this.uploadedFilesChangeTracker++
          this.uploadedFiles.set(uploadId, { name: event.name, size: event.size, status: 'done' })
        })
        .catch((e) => {
          console.error('Error uploading file ', event.name, ': ', e)
          this.uploadedFilesChangeTracker++
          this.uploadedFiles.set(uploadId, { name: event.name, size: event.size, status: 'error' })
        })
    },

    $_ocUploadingFileMessage(fileName) {
      const translated = this.$gettext('Uploading file "%{fileName}"')
      return this.$gettextInterpolate(translated, { fileName }, true)
    }
  }
}
</script>

<style lang="scss">
#files-drop-container {
  .oc-files-drop-drag-area {
    background: transparent;
    border: 1px dashed var(--oc-color-input-border);
    padding: var(--oc-space-xlarge);
  }
}
</style>
