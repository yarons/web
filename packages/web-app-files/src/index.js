import App from './App.vue'
import Favorites from './views/Favorites.vue'
import FilesDrop from './views/FilesDrop.vue'
import LocationPicker from './views/LocationPicker.vue'
import PrivateLink from './views/PrivateLink.vue'
import PublicFiles from './views/PublicFiles.vue'
import PublicLink from './views/PublicLink.vue'
import Personal from './views/Personal.vue'
import SharedResource from './views/shares/SharedResource.vue'
import SharedWithMe from './views/shares/SharedWithMe.vue'
import SharedWithOthers from './views/shares/SharedWithOthers.vue'
import SharedViaLink from './views/shares/SharedViaLink.vue'
import SpaceProject from './views/spaces/Project.vue'
import SpaceTrashbin from './views/spaces/Trashbin.vue'
import SpaceProjects from './views/spaces/Projects.vue'
import Trashbin from './views/Trashbin.vue'
import translations from '../l10n/translations.json'
import quickActions from './quickActions'
import store from './store'
import { FilterSearch, SDKSearch } from './search'
import { bus } from 'web-pkg/src/instance'
import { archiverService, thumbnailService, Registry } from './services'
import fileSideBars from './fileSideBars'
import { buildRoutes } from './router'
import get from 'lodash-es/get'

// dirty: importing view from other extension within project
import SearchResults from '../../web-app-search/src/views/List.vue'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appInfo = {
  name: $gettext('Files'),
  id: 'files',
  icon: 'resource-type-folder',
  isFileEditor: false,
  extensions: [],
  fileSideBars
}
const navItems = [
  {
    name(capabilities) {
      return capabilities.spaces?.enabled ? $gettext('Personal') : $gettext('All files')
    },
    icon: appInfo.icon,
    route: {
      path: `/${appInfo.id}/spaces/personal/home`
    }
  },
  {
    name: $gettext('Favorites'),
    icon: 'star',
    route: {
      path: `/${appInfo.id}/favorites`
    },
    enabled(capabilities) {
      return capabilities.files && capabilities.files.favorites
    }
  },
  {
    name: $gettext('Shares'),
    icon: 'share-forward',
    route: {
      path: `/${appInfo.id}/shares`
    }
  },
  {
    name: $gettext('Spaces'),
    icon: 'layout-grid',
    tag: $gettext('beta'),
    route: {
      path: `/${appInfo.id}/spaces/projects`
    },
    enabled(capabilities) {
      return capabilities.spaces && capabilities.spaces.enabled === true
    }
  },
  {
    name: $gettext('Deleted files'),
    icon: 'delete-bin-5',
    route: {
      path: `/${appInfo.id}/trash`
    },
    enabled(capabilities) {
      return capabilities.dav && capabilities.dav.trashbin === '1.0'
    }
  }
]

export default {
  appInfo,
  store,
  routes: buildRoutes({
    App,
    Favorites,
    Personal,
    FilesDrop,
    LocationPicker,
    PrivateLink,
    PublicFiles,
    PublicLink,
    SearchResults,
    SharedResource,
    SharedViaLink,
    SharedWithMe,
    SharedWithOthers,
    Spaces: {
      Project: SpaceProject,
      Projects: SpaceProjects,
      Trashbin: SpaceTrashbin
    },
    Trashbin
  }),
  navItems,
  quickActions,
  translations,
  ready({ router, store }) {
    Registry.filterSearch = new FilterSearch(store, router)
    Registry.sdkSearch = new SDKSearch(store, router)

    // when discussing the boot process of applications we need to implement a
    // registry that does not rely on call order, aka first register "on" and only after emit.
    bus.publish('app.search.register.provider', Registry.filterSearch)
    bus.publish('app.search.register.provider', Registry.sdkSearch)
  },
  userReady({ store }) {
    archiverService.initialize(
      store.getters.configuration.server || window.location.origin,
      get(store, 'getters.capabilities.files.archivers', [
        {
          enabled: true,
          version: '1.0.0',
          formats: ['tar', 'zip'],
          archiver_url: `${store.getters.configuration.server}index.php/apps/files/ajax/download.php`
        }
      ])
    )
    // FIXME: Remove mock data
    thumbnailService.initialize(
      get(store, 'getters.capabilities.files.thumbnail', {
        enabled: true,
        version: 'v0.1',
        supportedMimeTypes: ['image/png', 'image/jpg', 'image/jpeg', 'image/gif', 'text/plain']
      })
    )
  }
}
