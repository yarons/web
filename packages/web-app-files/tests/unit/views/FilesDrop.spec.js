import FilesDrop from '@files/src/views/FilesDrop.vue'
import { shallowMount } from '@vue/test-utils'
import GetTextPlugin from 'vue-gettext'
import { getStore, localVue } from './views.setup.js'
import { DavProperty } from 'web-pkg/src/constants'
import { linkRoleUploaderFolder } from '@files/src/helpers/share'
import VueCompositionAPI from '@vue/composition-api/dist/vue-composition-api'

localVue.use(VueCompositionAPI)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

localVue.prototype.$client.publicFiles = {
  PUBLIC_LINK_SHARE_OWNER: 'admin',
  // function is mocked because it should return a promise with a list of resources
  list: async () => [
    {
      getProperty: jest.fn((val) => {
        if (val === DavProperty.PublicLinkPermission) {
          return linkRoleUploaderFolder.bitmask(false)
        }
        return val
      })
    }
  ],
  // function takes token as an argument and is mocked because it should return some public link url
  getFileUrl: (token) => `http://some-url/${token}`,
  putFileContents: jest.fn()
}

const $route = {
  meta: {
    title: 'page route title'
  },
  params: {
    token: 'abc123def456'
  }
}

const selectors = {
  filesEmpty: '.files-empty',
  loadingHeader: '.oc-login-card-title'
}

const ocSpinnerStubSelector = 'oc-spinner-stub'

describe('FilesDrop', () => {
  it('should call "resolvePublicLink" method on wrapper mount', () => {
    const spyResolvePublicLink = jest.spyOn(FilesDrop.methods, 'resolvePublicLink')
    getShallowWrapper()

    expect(spyResolvePublicLink).toHaveBeenCalledTimes(1)
  })

  it('should show page title and configuration theme general slogan', () => {
    const wrapper = getShallowWrapper()

    expect(wrapper).toMatchSnapshot()
  })

  it('should show spinner with loading text if wrapper is loading', () => {
    const wrapper = getShallowWrapper({ loading: true })

    expect(wrapper.find(selectors.loadingHeader).exists()).toBeTruthy()
    expect(wrapper.find(ocSpinnerStubSelector).exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  describe('when "loading" is set to false', () => {
    const wrapper = getShallowWrapper()

    it('should not show spinner and loading header', () => {
      expect(wrapper.find(selectors.loadingHeader).exists()).toBeFalsy()
      expect(wrapper.find(ocSpinnerStubSelector).exists()).toBeFalsy()
    })

    it('should show share information title', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should show vue drop zone with given options', () => {
      const dropZone = wrapper.find('#files-drop-zone')
      expect(dropZone.exists()).toBeTruthy()
    })

    it('should show error message if only it has truthy value', () => {
      const wrapper = getShallowWrapper({
        loading: false,
        errorMessage: 'This is a test error message'
      })

      expect(wrapper).toMatchSnapshot()
    })
  })
})

function createStore(slogan = 'some slogan', davProperties = []) {
  return getStore({ slogan, davProperties })
}

function getShallowWrapper({ store = createStore(), loading = false, errorMessage = null } = {}) {
  return shallowMount(FilesDrop, {
    localVue,
    store,
    mocks: {
      $route,
      $router: {
        currentRoute: { name: 'some-route' },
        resolve: (r) => {
          return { href: r.name }
        },
        afterEach: jest.fn()
      },
      $uppyService: {
        $on: jest.fn(),
        useDropTarget: jest.fn(),
        useXhr: jest.fn()
      }
    },
    setup: () => {
      return {}
    },
    data() {
      return {
        loading: loading,
        errorMessage: errorMessage,
        share: {
          getProperty: (val) => val
        }
      }
    }
  })
}
