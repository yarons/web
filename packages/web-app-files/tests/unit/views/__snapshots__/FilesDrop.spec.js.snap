// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilesDrop should show page title and configuration theme general slogan 1`] = `
<div id="files-drop-container" class="oc-height-1-1 oc-flex oc-flex-column oc-flex-between">
  <h1 class="oc-invisible-sr">page route title</h1>
  <div class="oc-p oc-height-1-1">
    <div class="oc-flex oc-flex-column oc-flex-middle oc-height-1-1">
      <div class="oc-text-center oc-width-1-1 oc-width-xxlarge@m">
        <h2>admin shared this folder with you for uploading</h2>
        <file-upload-stub btnlabel="Drop files here to upload or click to select file" btnclass="" id="files-drop-zone" class="uk-flex uk-flex-middle uk-flex-center uk-placeholder"></file-upload-stub>
        <div id="previews" hidden="hidden"></div>
      </div>
      <!---->
    </div>
  </div>
  <div class="oc-text-center">
    <p>some slogan</p>
  </div>
</div>
`;

exports[`FilesDrop should show spinner with loading text if wrapper is loading 1`] = `
<div id="files-drop-container" class="oc-height-1-1 oc-flex oc-flex-column oc-flex-between">
  <h1 class="oc-invisible-sr">page route title</h1>
  <div class="oc-p oc-height-1-1">
    <div class="oc-flex oc-flex-column oc-flex-middle">
      <h2 class="oc-login-card-title">
        <translate-stub tag="span">Loading public linkâ€¦</translate-stub>
      </h2>
      <oc-spinner-stub arialabel="" size="medium" aria-hidden="true"></oc-spinner-stub>
    </div>
  </div>
  <div class="oc-text-center">
    <p>some slogan</p>
  </div>
</div>
`;

exports[`FilesDrop when "loading" is set to false should show error message if only it has truthy value 1`] = `
<div id="files-drop-container" class="oc-height-1-1 oc-flex oc-flex-column oc-flex-between">
  <h1 class="oc-invisible-sr">page route title</h1>
  <div class="oc-p oc-height-1-1">
    <div class="oc-flex oc-flex-column oc-flex-middle oc-height-1-1">
      <div class="oc-text-center oc-width-1-1 oc-width-xxlarge@m">
        <h2>admin shared this folder with you for uploading</h2>
        <file-upload-stub btnlabel="Drop files here to upload or click to select file" btnclass="" id="files-drop-zone" class="uk-flex uk-flex-middle uk-flex-center uk-placeholder"></file-upload-stub>
        <div id="previews" hidden="hidden"></div>
      </div>
      <div class="oc-text-center">
        <h2>
          <translate-stub tag="span">An error occurred while loading the public link</translate-stub>
        </h2>
        <p class="oc-m-rm">This is a test error message</p>
      </div>
    </div>
  </div>
  <div class="oc-text-center">
    <p>some slogan</p>
  </div>
</div>
`;

exports[`FilesDrop when "loading" is set to false should show share information title 1`] = `
<div id="files-drop-container" class="oc-height-1-1 oc-flex oc-flex-column oc-flex-between">
  <h1 class="oc-invisible-sr">page route title</h1>
  <div class="oc-p oc-height-1-1">
    <div class="oc-flex oc-flex-column oc-flex-middle oc-height-1-1">
      <div class="oc-text-center oc-width-1-1 oc-width-xxlarge@m">
        <h2>admin shared this folder with you for uploading</h2>
        <file-upload-stub btnlabel="Drop files here to upload or click to select file" btnclass="" id="files-drop-zone" class="uk-flex uk-flex-middle uk-flex-center uk-placeholder"></file-upload-stub>
        <div id="previews" hidden="hidden"></div>
      </div>
      <!---->
    </div>
  </div>
  <div class="oc-text-center">
    <p>some slogan</p>
  </div>
</div>
`;
